### DISCUSSION TITLE
Multiple matchers for grouping updates

### DISCUSSION THREAD

**Discussion by BadLiveware** [renovatebot/renovate#13644] [Request Help] [answered]
So I am trying to group major updates to both `docker` and `nuget` datasources because they are updated in sync. I have been able to group them together separately, 

for grouping nugets, `monorepo:dotnet`
```json
{
  "matchSourceUrlPrefixes": [
    "https://github.com/dotnet/aspnetcore",
    "https://github.com/dotnet/efcore",
    "https://github.com/dotnet/extensions",
    "https://github.com/dotnet/runtime"
  ]
}
```

for grouping container images, 
```json
{
  "packageRules": [
    {
      "groupName": "dotnet docker",
      "matchDatasources": ["docker"],
      "matchPackagePrefixes": ["mcr.microsoft.com/dotnet"],
    }
  ]
} 
```

So naturally I though I can combine these in order to group both together.
```json
{
  "packageRules": [
    {
      "groupName": "dotnet",
      "matchDatasources": ["docker", "nuget"],
      "matchPackagePrefixes": ["mcr.microsoft.com/dotnet"],
      "matchSourceUrlPrefixes": [
        "https://github.com/dotnet/aspnetcore",
        "https://github.com/dotnet/efcore",
        "https://github.com/dotnet/extensions",
        "https://github.com/dotnet/runtime",
      ],
    }
  ]
} 
```
This however causes the nuget packages to be grouped into one PR but the image tags in the Dockerfile are separate.

I did find a [previous discussion with the same topic ](https://github.com/renovatebot/renovate/discussions/8447) and using that answer did work to get both to group,
```json
{
  "packageRules": [
    {
      "groupName": "dotnet",
      "description": "Dotnet monorepo packages and container images",
      "matchUpdateTypes": ["major"],
      "matchDatasources": ["docker", "nuget"],
      "packagePatterns": [
        "dotnet$",
        "dotnet/sdk$",
        "dotnet/aspnet$",
        "dotnet/runtime$",
        "dotnet/nightly/runtime$",
        "dotnet/runtime-deps$",
        "dotnet/monitor$",
        "dotnet/samples$",
        "^Microsoft\\.Extensions.*$",
        "^Microsoft\\.EntityFrameworkCore.*$",
      ]
    },
  ]
} 
```
However, I do find it a bit cumbersome to match the nuget packages like this as there are independent packages that do not follow the same major versioning under the `Microsoft.` prefix and since there is a simpler way(`monorepo:dotnet`) I was wondering if I am able to combine this somehow?

---
Edit:
I have tried getting the docker matching using `matchSourceUrlPrefixes` but I never got it to work, it does group nuget but not container images
```json5
    {
      groupName: "dotnet",
      description: "Dotnet monorepo packages and container images",
      matchUpdateTypes: ["major"],
      matchDatasources: ["docker", "nuget"],
      matchSourceUrlPrefixes: [
        "https://github.com/dotnet/aspnetcore",
        "https://github.com/dotnet/efcore",
        "https://github.com/dotnet/extensions",
        "https://github.com/dotnet/runtime",

        // The source repo for the images
        "https://github.com/dotnet/dotnet-docker",
       // The dockerhub home for the images
        "https://hub.docker.com/_/microsoft-dotnet",
       // The prefix for the images in the Dockerfile
        "mcr.microsoft.com/dotnet",
       // The prefix for the images in the Dockerfile with protocol
        "https://mcr.microsoft.com/dotnet",
      ],
    },
```

**Comment by rarkins** [comment, marked as answer]
If you have multiple matches in single packageRule, then all of them must match. The only exception is that `matchPackageNames` and `matchPackagePatterns` are OR'd together.

If you want to match different types of stuff into one group then have more than one package rule, each with the same `groupName`.

**Reply by BadLiveware** [reply]
I suspected it had to do with the matchers being AND'd, setting the same `groupName` on multiple packageRules worked. Thank you for the quick reply and simple answer :)

**Reply by chkpnt** [reply]
What's about `matchPackagePrefixes`, is this OR'd, too?

**Reply by Gabriel-Ladzaretti** [reply]
@chkpnt yes,

`matchPackageNames`, `matchPackagePatterns` and `matchPackagePrefixes` are OR'd togather.
___
Other matchers with an OR logical relation between them are these -
`matchSourceUrls` & `matchSourceUrlPrefixes`

**Reply by gost-serb** [reply]
Would be nice to have this information in Renovate docs.
Especially considering that now `matchDepNames`, `matchDepPatterns`, `matchPackageNames`, `matchPackagePatterns`, `matchPackagePrefixes` are OR'd together. Correct me if I'm wrong.

**Reply by mediafreakch** [reply]
Definitely. Also how "extends" inside a package rule behaves together with one of `matchDepNames`, `matchPackageNames`, `matchPackagePatterns`...

**Reply by TWiStErRob** [reply]
An attempt at docs: https://github.com/renovatebot/renovate/pull/29264
